File: SampleApps /ObjC/SampleAppWakeUp/SampleAppWakeUp/Services/WakeupService/WakeupService.m
//
//  WakeupService.m
//  SampleAppWakeUp
//
//  Created by Akhil Rana on 24/11/23.
//

#import "WakeupService.h"
#import "Constants.h"

@interface WakeupService ()
@property (nonatomic, strong) CLLocationManager* locationManager;
@end

@implementation WakeupService

- (instancetype)init {
    self = [super init];
    if (self) {
        [self setUpLocationManager];
    }
    return self;
}

- (BOOL)isWakeUpEnabled {
    return _locationManager.monitoredRegions.count > 0;
}

- (void)setUpLocationManager {
    _locationManager = [[CLLocationManager alloc] init];
    _locationManager.delegate = self;
    _locationManager.pausesLocationUpdatesAutomatically = NO;
    _locationManager.allowsBackgroundLocationUpdates = YES;
    [_locationManager requestAlwaysAuthorization];
    [_locationManager requestWhenInUseAuthorization];
    
    _orgId = ORG_ID;
}

- (void)startMonitoringBeaconRegion {
    NSUUID *orgUUID = [[NSUUID alloc] initWithUUIDString:_orgId];
    if (!orgUUID) {
        NSLog(@">> Please enter a valid org_id");
        return;
    }
    
    NSMutableArray *beaconUUIDs = [NSMutableArray array];
    
    // Org Beacon
    [beaconUUIDs addObject:[orgUUID UUIDString]];
    
    // Org Cyclical Beacon
    NSString *shortUUID =  [self shortOrgUUIDString: [orgUUID UUIDString]];
    NSArray *beaconHexValues = @[@"a", @"b", @"c", @"d", @"e", @"f"];
    NSInteger numOfCyclicalBeacons = 16;
    
    for (NSInteger i = 0; i < numOfCyclicalBeacons; i++) {
        NSString *beaconUUID;
        if (i < 10) {
            beaconUUID = [NSString stringWithFormat:@"%@%ld", shortUUID, (long)i];
        } else if (i < 16) {
            beaconUUID = [NSString stringWithFormat:@"%@%@", shortUUID, beaconHexValues[i - 10]];
        } else {
            beaconUUID = [NSString stringWithFormat:@"%@%ld", shortUUID, (long)(i - 6)];
        }
        [beaconUUIDs addObject:beaconUUID];
    }
    
    for (NSInteger index = 0; index < beaconUUIDs.count; index++) {
        NSString *beaconUUIDString = beaconUUIDs[index];
        NSUUID *beaconUUID = [[NSUUID alloc] initWithUUIDString:beaconUUIDString];
        if (!beaconUUID) {
            return;
        }
        NSString *beaconName = [NSString stringWithFormat:@"Juniper_Mist_AP_Beacon_%ld", (long)index];
        CLBeaconRegion *beaconRegion = [self createBeaconRegionWithName:beaconName uuid:beaconUUID];
        [self.locationManager startMonitoringForRegion:beaconRegion];
        
        NSLog(@">> Start Monitoring Beacon = %@, proximityUUID = %@, major = %@ minor = %@", beaconRegion.identifier, beaconRegion.UUID.UUIDString, beaconRegion.major, beaconRegion.minor);
    }
    
}

- (void)stopMonitoringBeaconRegion {
    for (CLRegion *region in self.locationManager.monitoredRegions) {
        if ([region isKindOfClass:[CLBeaconRegion class]]) {
            CLBeaconRegion *beaconRegion = (CLBeaconRegion *)region;
            [self.locationManager stopMonitoringForRegion:beaconRegion];
                        
            NSLog(@">> Stop Monitoring Beacon = %@, proximityUUID = %@, major = %@ minor = %@", beaconRegion.identifier, beaconRegion.UUID.UUIDString, beaconRegion.major, beaconRegion.minor);
        }
    }
}

- (CLBeaconRegion *)createBeaconRegionWithName:(NSString *)name uuid:(NSUUID *)uuid {
    CLBeaconRegion *beaconRegion = [[CLBeaconRegion alloc] initWithUUID:uuid identifier:name];
    beaconRegion.notifyEntryStateOnDisplay = YES;
    return beaconRegion;
}

- (void)start {
    [self startMonitoringBeaconRegion];
}

- (void)stop {
    [self stopMonitoringBeaconRegion];
}

- (void)locationManager:(CLLocationManager *)manager didEnterRegion:(CLRegion *)region {
    [self.delegate didEnterRegion:region.identifier];
}

- (void)locationManager:(CLLocationManager *)manager didExitRegion:(CLRegion *)region {
    [self.delegate didExitRegion:region.identifier];
}

- (CLBeaconRegion *)createWithBeaconName:(NSString *)name uuid:(NSUUID *)uuid {
    CLBeaconRegion *beaconRegion = [[CLBeaconRegion alloc] initWithUUID:uuid identifier:name];
    beaconRegion.notifyOnEntry = YES;
    beaconRegion.notifyOnExit = YES;
    beaconRegion.notifyEntryStateOnDisplay = YES;
    return beaconRegion;
}

- (NSString *)shortOrgUUIDString:(NSString *)originalString {
    NSString *shortOrgId = [originalString substringToIndex:[originalString length] - 2];
    NSString *orgUUIDString = [NSString stringWithFormat:@"%@%@", shortOrgId, @"0"];
    return orgUUIDString;
}

@end

File: SampleApps/Swift/SampleBlueDotWithIndoorLocation/SampleBlueDotWithIndoorLocation/Services/WakeupService.swift
//
//  WakeupService.swift
//  SampleBlueDotWithIndoorLocation
//
//  Created by Rajesh Vishwakarma on 16/03/23.
//

import Foundation
import CoreLocation

protocol WakeUpService {
    var isWakeUpEnabled: Bool { get }
    var isWakeUpRuning: Bool { get }
    func start()
    func stop()
}

protocol WakeUpServiceDelegate: AnyObject {
    func didEnterRegion(region: String)
    func didExitRegion(region: String)
}

class RealWakeupService: NSObject {
    weak var delegate: WakeUpServiceDelegate?
    
    private lazy var locationManager = CLLocationManager()
    private let orgId = MistSDK.SDK.orgId
    
    var isWakeUpRuning: Bool = false
    
    override init() {
        super.init()
        setUpLocationManager()
    }
    
    private func setUpLocationManager() {
        locationManager.delegate = self
        locationManager.pausesLocationUpdatesAutomatically = false
        locationManager.allowsBackgroundLocationUpdates = true
        locationManager.requestAlwaysAuthorization()
        locationManager.requestWhenInUseAuthorization()
    }
    
    private func startMonitoringBeaconRegion() {
        guard let orgUUID =  UUID(uuidString: orgId) else {
            debugPrint(">> Please enter a valid org_id")
            return
        }
        
        var beaconUUIDs = [String]()
        
        // Org Beacon
        beaconUUIDs.append(orgUUID.uuidString)
        
        // Org Cyclical Beacon
        let shortUUID = orgUUID.uuidString.shortOrgUUIDString
        let beaconHexValues = ["a", "b", "c", "d", "e", "f"]
        let numOfCyclicalBeacons = 16
        
        for i in 0..<numOfCyclicalBeacons {
            if i < 10 {
                let beaconUUID = String(format: "%@%d", shortUUID, i)
                beaconUUIDs.append(beaconUUID)
            } else if i < 16 {
                let beaconUUID = String(format: "%@%@", shortUUID, beaconHexValues[i-10])
                beaconUUIDs.append(beaconUUID)
            } else {
                let beaconUUID = String(format: "%@%d", shortUUID, i-6)
                beaconUUIDs.append(beaconUUID)
            }
        }
        
        for (index, beaconUUID) in beaconUUIDs.enumerated() {
            guard let beaconUUID =  UUID(uuidString: beaconUUID) else { return }
            let beaconName = "Juniper_Mist_AP_Beacon_\(index)"
            let beaconRegion = CLBeaconRegion.create(with: beaconName, uuid: beaconUUID)
            locationManager.startMonitoring(for: beaconRegion)
            
            debugPrint(">> Start Monitoring Beacon = \(beaconRegion.identifier.description), proximityUUID = \(beaconRegion.uuid.uuidString), major = \(beaconRegion.major?.stringValue ?? "") minor = \(beaconRegion.minor?.stringValue ?? "")")
        }
        
        isWakeUpRuning = true
    }
    
    private func stopMonitoringBeaconRegion() {
        for region in locationManager.monitoredRegions {
            guard let beaconRegion = region as? CLBeaconRegion else { continue }
            locationManager.stopMonitoring(for: beaconRegion)
            
            isWakeUpRuning = false
            
            debugPrint(">> Stop Monitoring Beacon = \(beaconRegion.identifier.description), proximityUUID = \(beaconRegion.uuid.uuidString), major = \(beaconRegion.major?.stringValue ?? "") minor = \(beaconRegion.minor?.stringValue ?? "")")
        }
    }
}

extension RealWakeupService: CLLocationManagerDelegate {
    
    func locationManager(_ manager: CLLocationManager, didEnterRegion region: CLRegion) {
        delegate?.didEnterRegion(region: region.identifier)
    }
    
    func locationManager(_ manager: CLLocationManager, didExitRegion region: CLRegion) {
        delegate?.didExitRegion(region: region.identifier)
    }
}

extension RealWakeupService: WakeUpService {
    
    var isWakeUpEnabled:Bool {
        return locationManager.monitoredRegions.count > 0
    }
    
    func start() {
        startMonitoringBeaconRegion()
    }
    
    func stop() {
        stopMonitoringBeaconRegion()
    }
}
